quinn-proto/src/stream.rs

```diff
@@ -135,13 +135,25 @@ impl Streams {
 
     pub fn read(&mut self, id: StreamId, buf: &mut [u8]) -> Result<(usize, bool), ReadError> {
         let rs = self.get_recv_mut(id).ok_or(ReadError::UnknownStream)?;
-        Ok((rs.read(buf)?, rs.receiving_unknown_size()))
+        match rs.read(buf) {
+            Ok(len) => Ok((len, rs.receiving_unknown_size())),
+            Err(e) => Err(self.read_error_cleanup(id, e)),
+        }
     }
 
     pub fn read_unordered(&mut self, id: StreamId) -> Result<(Bytes, u64, bool), ReadError> {
         let rs = self.get_recv_mut(id).ok_or(ReadError::UnknownStream)?;
-        let (buf, len) = rs.read_unordered()?;
-        Ok((buf, len, rs.receiving_unknown_size()))
+        match rs.read_unordered() {
+            Ok((buf, len)) => Ok((buf, len, rs.receiving_unknown_size())),
+            Err(e) => Err(self.read_error_cleanup(id, e)),
+        }
+    }
+
+    fn read_error_cleanup(&mut self, id: StreamId, err: ReadError) -> ReadError {
+        if let ReadError::Finished | ReadError::Reset { .. } = err {
+            self.maybe_cleanup(id);
+        }
+        err
```

Comment:
Feels a bit weird to thread the `ReadError` through this, especially now that we're not being cheeky with `map_err` any more.

Reply:
Fair point, done.